File: Python/1448_Count Good Nodes in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode,max_value = float('-inf')) -> int:
        if not root:
            return 0            
        count = 1 if root.val >= max_value  else 0 
        max_value = max(max_value,root.val)
        return self.goodNodes(root.left,max_value) + self.goodNodes(root.right,max_value) + count
        
    
    def goodNodes_bfs(self, root: TreeNode) -> int:
        from collections import deque
        if not root:
            return 0
        queue = deque([(root,root.val)])     
        good = 1
        while queue:
            n = len(queue)
            for i in range(n): 
                node,val = queue.popleft()    
                if node.left:                                        
                    queue.append((node.left,max(val,node.left.val)))
                    if node.left.val >= val:
                        good += 1
                if node.right:
                    queue.append((node.right,max(val,node.right.val)))
                    if node.right.val >= val:
                        good += 1
        
        return good

File: C++/1448. Count Good Nodes in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode* root) {   
        return dfs(root,root->val);
    }
    
    int dfs(TreeNode* root,int maxValue){
        if (!root)
            return 0;        
        if (root->val >= maxValue){
            maxValue = root->val;
            return dfs(root->left,maxValue) + dfs(root->right,maxValue) + 1;
        }
        else{
            return dfs(root->left,maxValue) + dfs(root->right,maxValue);
        }  
    }   

    int goodNodes_bfs(TreeNode* root,int maxValue){
        if (!root)
            return 0; 
        
        int good = 1;
        queue<TreeNode*> q;
        q.push(root);
        
        while(!q.empty()){
            int size = q.size();
            for(int i = 0; i < size;i++){
                TreeNode *node = q.front(); 
                q.pop();
                if(node->left){
                    if(node->left->val >= node->val){
                        good++;                        
                    }      
                    else{
                        node->left->val = node->val;
                    }
                    q.push(node->left);
                }
                if(node->right){
                    if(node->right->val >= node->val){
                        good++;                        
                    }      
                    else{
                        node->right->val = node->val;
                    }
                    q.push(node->right);
                }                    
            }   
        }        
        return good;
    }     
    
};








